File: midian_of_two_sorted_arrays/go/midian_of_two_sorted_arrays.go
package midian

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {

	nums1Size := len(nums1)
	nums2Size := len(nums2)

	if nums1Size > nums2Size {
		return findMedianSortedArrays(nums2, nums1)
	}

	total := nums1Size + nums2Size

	if nums1Size == 0 {
		if total%2 == 1 {
			return float64(nums2[total/2])
		} else {
			return float64(nums2[total/2]+nums2[total/2-1]) / 2
		}
	}

	nums := make([]int, total)

	num1Counter := 0
	num2Counter := 0
	index := 0

	for i := 0; i < total; i++ {
		if nums1[num1Counter] < nums2[num2Counter] {
			nums[i] = nums1[num1Counter]
			num1Counter += 1
		} else if nums2[num2Counter] < nums1[num1Counter] {
			nums[i] = nums2[num2Counter]
			num2Counter += 1
		} else {
			nums[i] = nums1[num1Counter]
			num1Counter += 1
		}

		if num1Counter == nums1Size || num2Counter == nums2Size {
			index = i + 1
			break
		}
	}

	if num2Counter == nums2Size {
		for i := index; i < total; i++ {
			nums[i] = nums1[num1Counter]
			num1Counter += 1
		}
	} else {
		for i := index; i < total; i++ {
			nums[i] = nums2[num2Counter]
			num2Counter += 1
		}
	}

	if total%2 == 1 {
		return float64(nums[total/2])
	} else {
		return float64(nums[total/2]+nums[total/2-1]) / 2
	}
}

File: leet/midian_of_two_sorted_arrays/c/midian_of_two_sorted_arrays.c
#include <stdio.h>
#include <stdlib.h>

double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size) {

    if (nums1Size > nums2Size) {
        return findMedianSortedArrays(nums2, nums2Size, nums1, nums1Size);
    }

    int total = nums1Size+nums2Size;

    if (nums1Size == 0) {
        if (total % 2 == 1) {
            return nums2[total/2];
        } else {
            return (nums2[total/2] + nums2[total/2 - 1]) / (double)2;
        }
    }

    int nums[total];

    int num1Counter = 0;
    int num2Counter = 0;
    int index = 0;

    for (int i = 0; i < total; i++) {
        if (nums1[num1Counter] < nums2[num2Counter]){
            nums[i] = nums1[num1Counter++];
        } else if (nums2[num2Counter] < nums1[num1Counter]) {
            nums[i] = nums2[num2Counter++];
        } else {
            nums[i] = nums1[num1Counter++];
        }

        if (num1Counter == nums1Size || num2Counter == nums2Size) {
            index = i+1;
            break;
        }
    }

    if (num1Counter == nums1Size) {
        for (int i = index; i < total; i++) {
            nums[i] = nums2[num2Counter++];
        }
    } else {
        for (int i = index; i < total; i++) {
            nums[i] = nums1[num1Counter++];
        }
    }

    if (total % 2 == 1) {
        return nums[total/2];
    } else {
        return (nums[total/2] + nums[total/2 - 1]) / (double)2;
    }
}

int main() {
    printf("ok\n");

    int nums1[] = {3};
    int nums2[] = {-2, -1};

    printf("%f\n", findMedianSortedArrays(nums1, 1, nums2, 2));

    return 0;
}

